opendevin/action/base.py

```diff
@@ -0,0 +1,16 @@
+from dataclasses import dataclass
+
+from ..controller import AgentController
+
+@dataclass
+class Action:
+    pass
+
+    def run(self, controller: AgentController) -> str:
+        raise NotImplementedError
+
+    def to_dict(self):
+        return {
+            "action_type": self.__class__.__name__,
+            "args": self.__dict__
+        }
```

Comment:
Each action (optionally) takes in a `controller` and handles execution by itself.